spark/src/main/scala/geotrellis/spark/io/s3/S3RasterCatalog.scala

```diff
@@ -0,0 +1,122 @@
+            rootPath: String,
+            paramsConfig: DefaultParams[String] = BaseParams,
+            catalogConfig: S3RasterCatalogConfig = S3RasterCatalogConfig.DEFAULT)
+          (implicit sc: SparkContext): S3RasterCatalog = {    
+    
+    val attributeStore = new S3AttributeStore(catalogConfig.credentialsProvider, bucket, catalogConfig.layerDataDir)    
+    new S3RasterCatalog(bucket, rootPath, attributeStore, BaseParams, S3RasterCatalogConfig.DEFAULT)
+  }
+}
+
+class S3RasterCatalog(
+    bucket: String,
+    rootPath: String,
+    val attributeStore: S3AttributeStore,    
+    paramsConfig: DefaultParams[String],
+    catalogConfig: S3RasterCatalogConfig)
+  (implicit sc: SparkContext) {
+
+  def reader[K: RasterRDDReader: JsonFormat: ClassTag](): FilterableRasterRDDReader[K] =
+    new FilterableRasterRDDReader[K] {
+      def read(layerId: LayerId, filterSet: FilterSet[K]): RasterRDD[K] = {
+        val metaData = attributeStore.read[S3LayerMetaData](layerId, "metaData")
+        val keyBounds = attributeStore.read[KeyBounds[K]](layerId, "keyBounds")
+        val index = attributeStore.read[KeyIndex[K]](layerId, "keyIndex")
+        implicitly[RasterRDDReader[K]].read(catalogConfig.s3client, metaData, keyBounds, index)(layerId, filterSet)
+      }      
+    }
+
+  def writer[K: SpatialComponent: RasterRDDWriter: Boundable: JsonFormat: ClassTag](keyIndexMethod: KeyIndexMethod[K]): Writer[LayerId, RasterRDD[K]] =
+    writer[K](keyIndexMethod, clobber = true)
+
+  def writer[K: SpatialComponent: RasterRDDWriter: Boundable: JsonFormat: ClassTag](keyIndexMethod: KeyIndexMethod[K], subDir: String): Writer[LayerId, RasterRDD[K]] =
+    writer[K](keyIndexMethod, subDir, clobber = true)
+
+  def writer[K: SpatialComponent: RasterRDDWriter: Boundable: JsonFormat: ClassTag](keyIndexMethod: KeyIndexMethod[K], clobber: Boolean): Writer[LayerId, RasterRDD[K]] =
+    new Writer[LayerId, RasterRDD[K]] {
+      def write(layerId: LayerId, rdd: RasterRDD[K]): Unit = {
+        writer[K](keyIndexMethod, paramsConfig.paramsFor[K](layerId).getOrElse(""), true).write(layerId, rdd)
+      }
+    }
+  
+  def writer[K: SpatialComponent: RasterRDDWriter: Boundable: JsonFormat: ClassTag](keyIndexMethod: KeyIndexMethod[K], subDir: String, clobber: Boolean): Writer[LayerId, RasterRDD[K]] =
+    new Writer[LayerId, RasterRDD[K]] {
+      def write(layerId: LayerId, rdd: RasterRDD[K]): Unit = {
+        rdd.persist()
+
+        val layerPath = 
+          if (subDir != "")
+            s"${rootPath}/${subDir}/${catalogConfig.layerDataDir(layerId)}"
+          else
+            s"${rootPath}/${catalogConfig.layerDataDir(layerId)}"
+
+        val md = S3LayerMetaData(
+            layerId = layerId,
+            keyClass = classTag[K].toString,
+            rasterMetaData = rdd.metaData,
+            bucket = bucket,
+            key = layerPath)
+
+        // Note: this is a waste to do for spatial raster, we can derive them easily from RasterMetaData
```

Comment:
This is true, so shouldn't this method be implemented by the particular writer (for spatial or spacetime?). I see this writer is generic, but this would be a good use case for following the `SpatialRasterRDDWriterProvider` pattern like the other catalogs.